src/Umbraco.Core/Persistence/Repositories/ContentRepository.cs

```diff
@@ -263,6 +263,27 @@ public void RebuildXmlStructures(Func<IContent, XElement> serializer, int groupS
                 }
                 baseId = xmlItems[xmlItems.Count - 1].NodeId;
             }
+
+            //now delete the items that shouldn't be there
+            var sqlAllIds = translate(0, GetBaseQuery(BaseQueryType.Ids));
+            var allContentIds = Database.Fetch<int>(sqlAllIds);
+            var docObjectType = Guid.Parse(Constants.ObjectTypes.Document);
+            var xmlIdsQuery = new Sql()
+                .Select("DISTINCT cmsContentXml.nodeId")
+                .From<ContentXmlDto>(SqlSyntax)
+                .InnerJoin<NodeDto>(SqlSyntax)
+                .On<ContentXmlDto, NodeDto>(SqlSyntax, left => left.NodeId, right => right.NodeId)
+                .Where<NodeDto>(dto => dto.NodeObjectType == docObjectType, SqlSyntax);
+            if (contentTypeIdsA.Length > 0)
+            {
+                xmlIdsQuery.WhereIn<ContentDto>(x => x.ContentTypeId, contentTypeIdsA, SqlSyntax);
+            }
+
+            var allXmlIds = Database.Fetch<int>(xmlIdsQuery);
+
+            var toRemove = allXmlIds.Except(allContentIds).ToArray();
+            if (toRemove.Length > 0)
+                Database.Execute("DELETE FROM cmsContentXml WHERE nodeId IN (@ids)", new { ids = toRemove });

```

Comment:
Hi @Shazwazza and @zpqrtbnk 

Suggestion if it is possible with your "QueryBuilder"

Why don't you do one single delete query. That way you only have one call to the DB and the DB-server handles everything?

But only if the querybuilder can make something like this.

DELETE cmsContentXml
FROM cmsContentXml
INNER JOIN cmsNode ON cmsContentXml.nodeId = cmsNode.nodeId
WHERE cmdNode.NodeObjectType = {docObjectType}

Don't know if your "WhereIn" (line 279) is possible to add the extra where-expression to the example.

This is just a suggestion, so the C# code doesn't have to handle the content-ids and call the db more than needed. And you don't have to think of doing "InGroupsOf".